### PR TITLE
Speed up FA Backward pass in GPU via parallelizing sequence dimension

### PR DESCRIPTION
**Pull request by kelvin-zou**
Following the examples in https://github.com/triton-lang/triton/blob/main/python/tutorials/06-fused-attention.py and https://github.com/jax-ml/jax-triton/pull/177/files and split the dkdv and dq in backward for better seq parallel .

With this change, AxLearn Triton kernel now is much faster than before for 2k+ sequence length, and it can deliver end2end 15% speedup over old triton kernel which makes it a more palatable to use over CuDNN kernel. The exact changes in this PR are:
1. Adding seq parallel in backward pass.
2. Tune num_stage/num_warps to better suit common use case without hitting SRAM OOM.
3. Updated the benchmark test a bit to better match real prod setting.

Benchmark perf numbers: 
Set up: Batch size : 2,  Num of Heads: 16, Hidden Dim Per Head: 128
|run id |   seq_len |      XLA       |  AxLearn Old FA | Axlearn New FA |     Pallas    |  CuDNN   |
| --- | --- | --- | --- | --- | --- | --- | 
|0         |  1024.0    | 1.520580   | 1.360455       |  1.405664 | 1.334541  | 1.603672|
|1          |  2048.0   | 1.432180   | 1.827646        |  1.310446 |1.269701    |1.612260|
|2         | 4096.0    | 4.451288   | 7.259165        |0.694067   |0.639129   | 1.431808|
|3         | 8192.0     | 16.915710  | 28.574991     |  1.633464 |1.420990   | 1.052758|
|4         | 16384.0   | 61.869404|  112.639290  | 4.858109   |4.104518   | 2.245235|

### CODE REVIEW

File: axlearn/common/flash_attention/gpu_attention_benchmark.py [hunk 1]
```diff
@@ -82,24 +82,24 @@
...
 def _perf_report(prefix: str):
     # 128 is the most common value for per_head_dim.
-    batch_size, num_heads, seq_len, per_head_dim = 2, 32, 2048, 128
+    batch_size, num_heads, seq_len, per_head_dim = 2, 16, 8192, 128
 
     # Vary batch size for fixed heads and seq length.
     batch_size_bench = triton.testing.Benchmark(
         x_names=["batch_size"],
-        x_vals=[2, 4, 8, 16],
+        x_vals=[1, 2, 4],
```
[the remark this answers is not in the document]
Comment (kelvin-zou) [reply axlearn/common/flash_attention/gpu_attention_benchmark.py]: Done

File: axlearn/common/flash_attention/gpu_attention.py [hunk 2]
```diff
@@ -536,71 +539,123 @@ def _mha_backward_kernel(
...
+    # Load a block of K and V of size (block_k, block_d).
+    # Iterate through Q in chunks of (block_q, block_d) to accumulate dK and dV.
+    start_k = pl.program_id(2)
+    slice_k = pl.ds(start_k * block_k, block_k)
+    dv = jnp.zeros([block_k, block_d], dtype=jnp.float32)
+    dk = jnp.zeros([block_k, block_d], dtype=jnp.float32)
+    k = pl.load(k_ref, (slice_k, slice(None)))
+    v = pl.load(v_ref, (slice_k, slice(None)))
+    span_k = start_k * block_k + jnp.arange(block_k)
+    kv_segment_ids = None if s_ref is None else pl.load(s_ref, (slice_k))
```
[the remark this answers is not in the document]
Comment (kelvin-zou) [reply axlearn/common/flash_attention/gpu_attention.py]: Done

File: axlearn/common/flash_attention/gpu_attention.py [hunk 3]
```diff
@@ -536,71 +539,123 @@ def _mha_backward_kernel(
     del out_ref, l_ref  # Not needed
```
[the remark this answers is not in the document]
Comment (kelvin-zou) [reply axlearn/common/flash_attention/gpu_attention.py:544]: Done.

**Review by markblee — APPROVED** [review]
